lib/contents.php

```diff
@@ -1,244 +1,133 @@
+	curl_setopt($ch, CURLOPT_USERAGENT, $config['useragent']);
+	curl_setopt($ch, CURLOPT_TIMEOUT, $config['timeout']);
 	curl_setopt($ch, CURLOPT_ENCODING, '');
 	curl_setopt($ch, CURLOPT_PROTOCOLS, CURLPROTO_HTTP | CURLPROTO_HTTPS);
-
-	if(is_array($opts) && count($opts) !== 0) {
-
-		Debug::log('Setting options: ' . json_encode($opts));
-
-		foreach($opts as $key => $value) {
-			curl_setopt($ch, $key, $value);
-		}
+	if($config['proxy']) {
+		curl_setopt($ch, CURLOPT_PROXY, $config['proxy']);
 	}
-
-	if(defined('PROXY_URL') && !defined('NOPROXY')) {
-
-		Debug::log('Setting proxy url: ' . PROXY_URL);
-		curl_setopt($ch, CURLOPT_PROXY, PROXY_URL);
-
+	foreach($config['curl_options'] as $key => $value) {
+		curl_setopt($ch, $key, $value);
 	}
-
-	// We always want the response header as part of the data!
-	curl_setopt($ch, CURLOPT_HEADER, true);
-
-	// Build "If-Modified-Since" header
-	if(!Debug::isEnabled() && $time = $cache->getTime()) { // Skip if cache file doesn't exist
-		Debug::log('Adding If-Modified-Since');
-		curl_setopt($ch, CURLOPT_TIMEVALUE, $time);
+	if ($config['if_not_modified_since']) {
+		curl_setopt($ch, CURLOPT_TIMEVALUE, $config['if_not_modified_since']);
 		curl_setopt($ch, CURLOPT_TIMECONDITION, CURL_TIMECOND_IFMODSINCE);
 	}
 
-	// Enables logging for the outgoing header
-	curl_setopt($ch, CURLINFO_HEADER_OUT, true);
-
+	$responseHeaders = [];
+	curl_setopt($ch, CURLOPT_HEADERFUNCTION, function ($ch, $rawHeader) use (&$responseHeaders) {
+		$len = strlen($rawHeader);
+		if (preg_match('#^HTTP/(2|1.1|1.0)#', $rawHeader) || $rawHeader === "\r\n") {
+			return $len;
+		}
+		$header = explode(':', $rawHeader);
+		if (count($header) === 1) {
+			return $len;
+		}
+		$name = mb_strtolower(trim($header[0]));
+		$value = trim(implode(':', array_slice($header, 1)));
+		if (!isset($responseHeaders[$name])) {
+			$responseHeaders[$name] = [];
+		}
+		$responseHeaders[$name][] = $value;
+		return $len;
+	});
 	$data = curl_exec($ch);
-	$errorCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
-
-	$curlError = curl_error($ch);
-	$curlErrno = curl_errno($ch);
-	$curlInfo = curl_getinfo($ch);
-
-	Debug::log('Outgoing header: ' . json_encode($curlInfo));
-
-	if($data === false)
-		Debug::log('Cant\'t download ' . $url . ' cUrl error: ' . $curlError . ' (' . $curlErrno . ')');
-
-	$headerSize = curl_getinfo($ch, CURLINFO_HEADER_SIZE);
-	$header = substr($data, 0, $headerSize);
-	$retVal['header'] = $header;
-
-	Debug::log('Response header: ' . $header);
-
-	$headers = parseResponseHeader($header);
-	$finalHeader = end($headers);
-
-	curl_close($ch);
-
-	$finalHeader = array_change_key_case($finalHeader, CASE_LOWER);
-
-	switch($errorCode) {
-		case 200: // Contents OK
-		case 201: // Contents Created
-		case 202: // Contents Accepted
-			Debug::log('New contents received');
-			$data = substr($data, $headerSize);
-			// Disable caching if the server responds with "Cache-Control: no-cache"
-			// or "Cache-Control: no-store"
```

Comment:
Less code is better. If you want to debug http requests you can add back manually what you want. I also have a later plan for better logging instead of using the debug flag. We should log more, regardless of the debug flag.

Reply:
I am referring to the comments, not the logging. When you remove comments, the code becomes a wall that is difficult to read if you are not already familiar with it.